### PR TITLE
Feature: MSE metrics inside/outside regional domain for stretched grid models (KNMI)

### PR DESCRIPTION
**Pull request by jswijnands**
_Description of functionality_
Four new metrics have been added for stretched grid models. The first two metrics compute the weighted MSE (i) only inside the regional domain and (ii) only outside the regional domain. The former will allow for comparison of LAM and stretched grid models in the future. Further, two additional metrics have been added that compute the contributions of these components to the overall weighted MSE. This enables monitoring which of the two is the main driver of the training process, as it indicates whether we are mainly optimising the loss inside or outside the regional domain.

_Configuration options_
Both features can be turned on or off in the config options by specifying the following stretched grid metrics (sg_metrics) in the diagnostics config:
- wmse_per_region (True/False)
- wmse_contributions (True/False)

_Outputs_
Plots of these metrics are generated in MLFlow, specified as  ‘wmse_inside_lam_epoch’, ‘wmse_outside_lam_epoch’, ‘wmse_inside_lam_contribution_epoch’, and ‘wmse_outside_lam_contribution_epoch’.

MLFlow plots w-MSE splitting_default_area_weighting

_Further comments_
- Note that this PR is dependent on a PR that provides the stretched_grid.yaml graph config.
- This PR has also been tested together with the PR on custom domain weighting, to check that these features interact as expected. 
- A CERRA config has been added for testing purposes, but does not have to be merged into anemoi.

### CODE REVIEW

**Comment by FussyDuck** [comment]
[CLA assistant check](https://bol-claassistant.ecmwf.int:443/ecmwf/anemoi-training?pullRequest=130) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://bol-claassistant.ecmwf.int:443/ecmwf/anemoi-training?pullRequest=130) before we can accept your contribution.<br/>**1** out of **3** committers have signed the CLA.<br/><br/>:white_check_mark: jswijnands<br/>:x: Sophie Buurman<br/>:x: havardhhaugen<br/><hr/>**Sophie Buurman** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://bol-claassistant.ecmwf.int:443/check/ecmwf/anemoi-training?pullRequest=130) it.</sub>

File: src/anemoi/training/losses/mse.py
Comment (HCookie) [inline src/anemoi/training/losses/mse.py:1]: Additonally, the metrics have also been refactored. I have a couple comments, and will leave them here, but if needed we can discuss on a call.

File: src/anemoi/training/losses/mse.py [hunk 1]
```diff
@@ -89,3 +89,109 @@ def forward(
         # keep last dimension (variables) when summing weights
         out /= self.sum_function(self.weights[..., None].expand_as(out), axis=(0, 1, 2))
         return self.sum_function(out, axis=(0, 1, 2))
+
+
+class WeightedMSELossStretchedGrid(nn.Module):
```
Comment (HCookie) [inline src/anemoi/training/losses/mse.py:94]: This should subclass from the base `WeightedMSE` or `BaseWeighted`.

File: src/anemoi/training/losses/mse.py [hunk 2]
```diff
@@ -89,3 +89,109 @@ def forward(
...
+    or as the contribution to the overall loss.
+    """
+
+    def __init__(
+        self,
+        node_weights: torch.Tensor,
+        mask: torch.Tensor,
+        inside_lam: bool = True,
+        wmse_contribution: bool = False,
+        data_variances: torch.Tensor | None = None,
```
Comment (HCookie) [inline src/anemoi/training/losses/mse.py:107]: This has been removed in favour of the `scalars`. These provide a way to use arbitary scalars on the loss calculation.
I think it may be possible to abstract `mask` into these scalars from the forecaster as well

File: src/anemoi/training/losses/mse.py [hunk 3]
```diff
@@ -89,3 +89,109 @@ def forward(
...
+        self.sum_function = torch.nansum if ignore_nans else torch.sum
+
+        self.inside_lam = inside_lam
+        self.wmse_contribution = wmse_contribution
+        self.register_buffer("weights", node_weights, persistent=True)
+        self.register_buffer("weights_inside_lam", node_weights[mask], persistent=True)
+        self.register_buffer("weights_outside_lam", node_weights[~mask], persistent=True)
+        self.register_buffer("mask", mask, persistent=True)
+        if data_variances is not None:
+            self.register_buffer("ivar", data_variances, persistent=True)
```
Comment (HCookie) [inline src/anemoi/training/losses/mse.py:139]: By subclass correctly, some of the init fluff can be removed

File: src/anemoi/training/losses/mse.py [hunk 4]
```diff
@@ -89,3 +89,109 @@ def forward(
...
+                out /= self.sum_function(weights_selected.expand_as(out))
+            return self.sum_function(out)
+
+        # Weight by area
+        out = out * weights_selected[..., None].expand_as(out)
+        if self.wmse_contribution:
+            out /= self.sum_function(self.weights[..., None].expand_as(full_out_dims))
+        else:
+            out /= self.sum_function(weights_selected[..., None].expand_as(out))
+        return self.sum_function(out, axis=(0, 1, 2))
```
Comment (HCookie) [inline src/anemoi/training/losses/mse.py:197]: If the `mask` is to be used within the `scalars`, this will need to be refactored. I am happy to propose a refactor if you can provide some expected results to test against.

File: src/anemoi/training/train/forecaster.py [hunk 5]
```diff
@@ -99,6 +101,47 @@ def __init__(
...
+                    wmse_contribution=True,
+                    data_variances=loss_scaling,
+                )
+                self.sg_metrics["wmse_outside_lam_contribution_epoch"] = WeightedMSELossStretchedGrid(
+                    node_weights=self.loss_weights,
+                    mask=self.mask,
+                    inside_lam=False,
+                    wmse_contribution=True,
+                    data_variances=loss_scaling,
+                )
```
Comment (HCookie) [inline src/anemoi/training/train/forecaster.py:141]: This could be simplified into a `permute`

**Review by HCookie — CHANGES_REQUESTED** [review]

File: src/anemoi/training/config/stretched_grid_cerra.yaml
[the remark this answers is not in the document]
Comment (jswijnands) [reply src/anemoi/training/config/stretched_grid_cerra.yaml:1]: See the final line of the description: the reason the CERRA config was included is in case you'd like to do any testing. There is no need to merge this file into anemoi-training. I can remove it from the PR if that makes it more clear.

File: src/anemoi/training/diagnostics/callbacks/__init__.py [hunk 6]
```diff
@@ -721,6 +739,43 @@ def _plot(
...
+                    output_tensor[rollout_step, :, ~pl_module.mask, :],
+                )
+                self._output_figure(
+                    logger,
+                    fig_lam_outside,
+                    epoch=epoch,
+                    tag=f"pred_val_sample_outside_lam_rstep{rollout_step:02d}_batch{batch_idx:04d}_rank0",
+                    exp_log_tag=f"val_pred_sample_outside_lam_rstep{rollout_step:02d}_rank{local_rank:01d}",
+                )
+
```
Comment (HCookie) [inline src/anemoi/training/diagnostics/callbacks/__init__.py:778]: It may be good to have these plotting calls in another callback specific to StretchedGrid, that way a user can choose to toggle on or off. Then an assert can be raised when attempting to plot without the appropriate config.

File: src/anemoi/training/losses/mse.py (hunk 4 above)
Comment (HCookie) [reply src/anemoi/training/losses/mse.py:197]: Some changes are being made to the scalar that will assist in the refactor of this code. https://github.com/ecmwf/anemoi-training/pull/137
